src/modules/aws_ec2/mappers/instance_block_device_mapping.ts

```diff
@@ -0,0 +1,264 @@
+import { WaiterState } from '@aws-sdk/util-waiter';
+
+import { AwsEc2Module } from '..';
+import { AWS, crudBuilder2 } from '../../../services/aws_macros';
+import { Context, Crud2, MapperBase } from '../../interfaces';
+import { GeneralPurposeVolume, Instance, InstanceBlockDeviceMapping, VolumeState } from '../entity';
+
+export class InstanceBlockDeviceMappingMapper extends MapperBase<InstanceBlockDeviceMapping> {
+  module: AwsEc2Module;
+  entity = InstanceBlockDeviceMapping;
+  equals = (a: InstanceBlockDeviceMapping, b: InstanceBlockDeviceMapping) => {
+    return Object.is(a.deviceName, b.deviceName) && Object.is(a.deleteOnTermination, b.deleteOnTermination);
+  };
+
+  attachVolumeInternal = crudBuilder2<EC2, 'attachVolume'>(
+    'attachVolume',
+    (VolumeId, InstanceId, Device) => ({
+      VolumeId,
+      InstanceId,
+      Device,
+    }),
+  );
+
+  modifyBlockDeviceMapping = crudBuilder2<EC2, 'modifyInstanceAttribute'>(
+    'modifyInstanceAttribute',
+    (InstanceId, BlockDeviceMappings) => ({
+      InstanceId,
+      BlockDeviceMappings,
+    }),
+  );
+
+  waitUntilAvailable(client: EC2, volumeId: string) {
+    return this.module.generalPurposeVolume.volumeWaiter(client, volumeId, (vol: AWSVolume | undefined) => {
+      // If state is not 'in-use' retry
+      if (!Object.is(vol?.State, VolumeState.AVAILABLE)) {
+        return { state: WaiterState.RETRY };
+      }
+      return { state: WaiterState.SUCCESS };
+    });
+  }
+
+  detachVolumeInternal = crudBuilder2<EC2, 'detachVolume'>('detachVolume', VolumeId => ({ VolumeId }));
+
+  detachVolume = async (client: EC2, VolumeId: string) => {
+    await this.detachVolumeInternal(client, VolumeId);
+    await this.waitUntilAvailable(client, VolumeId);
+  };
+
+  waitUntilInUse(client: EC2, volumeId: string) {
+    return this.module.generalPurposeVolume.volumeWaiter(client, volumeId, (vol: AWSVolume | undefined) => {
+      // If state is not 'in-use' retry
+      if (!Object.is(vol?.State, VolumeState.IN_USE)) {
+        return { state: WaiterState.RETRY };
+      }
+      return { state: WaiterState.SUCCESS };
+    });
+  }
+
+  attachVolume = async (client: EC2, VolumeId: string, InstanceId: string, Device: string) => {
+    const result = await this.attachVolumeInternal(client, VolumeId, InstanceId, Device);
+    if (result) await this.waitUntilInUse(client, VolumeId);
+    else return false;
+    return true;
+  };
+
+  cloud: Crud2<InstanceBlockDeviceMapping> = new Crud2({
+    create: async (es: InstanceBlockDeviceMapping[], ctx: Context) => {
+      const out: InstanceBlockDeviceMapping[] = [];
+      for (const e of es) {
+        // read instance details
+        const instance: Instance = await ctx.orm.findOne(Instance, {
+          id: e.instanceId,
+        });
+
+        // if we have different regions we may be in a multi-region update situation, just skip
+        if (e.region !== instance.region) continue;
```

Comment:
Can we actually get into this situation? Or is this just a safeguard in the code?

Because I think we can prevent this sort of craziness by interlocking `region` joins where the `InstanceBlockDeviceMapping` has a `region` column that is FK joined to both the Instance and GPV entities.

Reply:
it actually happened when i update an instance moving to a different region. So region is updated at instance but mapping still points to the older region, until actually things get reconciled. There is a region field on the instance_block_device_mapping but is actually independent on the instance/gpv, and i just enforce it by code. How would you enforce that from the database itself?

Reply:
So moving an instance to another region becomes *very* tough when you do this because now it will fail when the FK relation on this `InstanceBlockDeviceMapping` table still has the old region.

One example of where we do it is with [`Subnet`](https://github.com/iasql/iasql/blob/main/src/modules/aws_vpc/entity/subnet.ts#L37-L52)s joining to the Availability Zone. You could follow a pattern like that including the region in the FK join of `InstanceBlockDeviceMapping` to both `Instance` and `GeneralPurposeVolume`

It would, for the most part, make mutating the `instance`'s region impossible from the FK constraints because it's volumes are in the same region. There would be a trick in a transaction to first delete the `InstanceBlockDeviceMapping` record(s), then change the `region` in GPV and Instance, then re-insert the `InstanceBlockDeviceMapping` record(s), *then* run `iasql_commit`, but that could reasonably be treated by the engine as wholly separate entities and not migrating the volume data between regions would be reasonable.